### ISSUE TITLE
DataStream.flatMap causes stream to end early

### ISSUE THREAD

**Issue by maxharlow** [signicode/scramjet#99]
When I use `flatMap` the stream seems to end prematurely in some circumstances. For example:

```
const Scramjet = require('scramjet')

const array = Array.from(new Array(50).keys())
Scramjet.DataStream.from(array)
    .flatMap(x => [x, x])
    .each(console.log)
```

I'd expect this to print out every number from 0 to 49 twice. Instead it consistently stops at 14. If I change the `flatMap` line to be `x => [x]` it runs all the way through as expected. Changing it to `x => [x, x, x]` makes it run through to 24.

I'm using Scramjet v4.35.10 and Node v15.5.1.

**Comment by maxharlow** [comment]
Downgrading to Scramjet v4.23.2 seems to fix the problem

**Comment by MichalCz** [comment]
Thanks for the info, will check today and report back

**Comment by MichalCz** [comment]
BTW, can you check if you get to see the `end` event?

**Comment by maxharlow** [comment]
Brilliant, thanks. How do I check for the `end` event?

**Comment by MichalCz** [comment]
on('end', /* some callback here */)

**Comment by maxharlow** [comment]
Just checked, no end event

**Comment by MichalCz** [comment]
Ok. Then it should be an easy fix.

M.

**Comment by MichalCz** [comment]
Max, I've found a fix, but I do need to investigate a bit and clean up the code after all this.

I can commit to Monday, I hope this works for you.

**Comment by maxharlow** [comment]
That's great! No rush, I can use the older version for now

**Comment by MichalCz** [comment]
@maxharlow this is fixed in `scramjet@4.35.11`
